cmake/Config.cmake

```diff
@@ -8,15 +8,15 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
     # detect the architecture (note: this test won't work for cross-compilation)
     include(CheckTypeSize)
     check_type_size(void* SIZEOF_VOID_PTR)
-    if("${SIZEOF_VOID_PTR}" STREQUAL "4")
+    if(${SIZEOF_VOID_PTR} STREQUAL "4")
```

Comment:
Is that really correct? (Any references?)
I mainly ask because I remember that this was a bit of a strange CMake thing with quotes and STREQUAL (reminds me of Batch files...).

Reply:
I accidentally found the corresponding pull request #1102. I'm not sure if I should write comments on closed tickets or merged commits but maybe someone want to lern from it.

The proposed change does not solve the problem. After applying this change, the condition would even be true for `SIZEOF_VOID_PTR == "1;OR;ABC"` since it is unrolled to `if(1 OR ABC STREQUAL 4`).

However, the pull request was right that there existed a flaw which is not been fixed yet. CMake is quite strange in some cases and there does not seem to be an proper solution. The statement `X STREQUAL Y` does actually check whether there is a variable `X` or `Y`. In this case, it will use the content of the variable, otherwise it would just use `X` or `Y`. It is described in the [documentation](https://cmake.org/cmake/help/v2.8.11/cmake.html#command:if). This means you could use `if(SIZEOF_VOID_PTR STREQUAL "4")` (without `${}`) to avoid double-dereferencing the variable. However, when you would compare it to a string instead of a number (`4`), that string might be dereferenced as well. `MATCHES` does not dereference the right operand but it does not seem to support exact matches. So, there does not seem to be a proper solution without adding extra lines for every if.

Luckily, there has been added a proper solution through [policy CMP0054](https://cmake.org/cmake/help/v3.1/policy/CMP0054.html) in CMake 3.1.